### PR TITLE
New Conversion functions with integers as inputs

### PR DESCRIPTION
**Pull request by buscandoaverroes**
1. developed `ConvertExtendedSpatialIDsToQuadkeysAndVerticalIDsV2` (name to be decided later)
- uses newly developed `convertVerticalIndex` to convert the vertical index
- `calculateMinVerticalIndex` returns the index that satisfies the minimum equality condition between conversion parameters
- `returnAltitudesOfVerticalIndex` returns the top and bottom altitudes of a given index
- `calculateVerticalResolution` returns the number of indices in a given altitude range
- `calculateVoxelHeight` returns the height of each voxel
- passes package tests

### CODE REVIEW

File: transform/convert_quadkey_and_Vertical_id.go [hunk 1]
```diff
@@ -576,6 +662,122 @@ func convertVerticallIDToBit(vZoom int64, vIndex int64, outputZoom int64, maxHei
...
+//	outputZoom: the zoomLevel of the outputIndex, before using zoomScalar (net outputZoom = outputZoom+zoomScalar)
+//
+//	zoomScalar: a scalar to add to outputZoom (net outputZoom = outputZoom+zoomScalar)
+//
+//	offset: an integer to shift the height of the inputIndex up or down. A positive offset means the transformed index is higher in altitude than that of the input.
+//
+// output:
+//
+//	[]int64 list of vertical index(es), error
+func convertVerticalIndex(inputIndex int64, inputZoom int64, outputZoom int64, altitudeRangeScalar int64, indexOffset int64) ([]int64, error) {
```
Comment (HarutakaMatsumoto) [inline transform/convert_quadkey_and_Vertical_id.go:682]: Rename altitudeRangeScalar to zoomOffset.
If you remove outputZoom like the following one, the return value will no longer be a slice and will be very simple.
Don't you remove outputZoom?
```
func convertVerticalIndex(inputIndex int64, zoomOffset int64, indexOffset int64) int64 {
    return inputIndex >> zoomOffset + indexOffset
}
```

File: transform/convert_quadkey_and_Vertical_id.go [hunk 2]
```diff
@@ -342,6 +348,86 @@ func ConvertExtendedSpatialIDsToQuadkeysAndVerticalIDs(extendedSpatialIDs []stri
 	return extendedSpatialIDToQuadkeyAndVerticalID, nil
 }
 
+func ConvertExtendedSpatialIDsToQuadkeysAndVerticalIDsV2(extendedSpatialIDs []string, outputHZoom int64, outputVZoom int64, altitudeRangeScalar int64, verticalIndexOffset int64) ([]*object.FromExtendedSpatialIDToQuadkeyAndVerticalID, error) {
```
Comment (HarutakaMatsumoto) [inline transform/convert_quadkey_and_Vertical_id.go:351]: Rename ConvertExtendedSpatialIDsToQuadkeysAndVerticalIDsV2 to ConvertExtendedSpatialIDsToQuadkeysAndAltitudekeys.
Define a new type FromExtendedSpatialIDToQuadkeyAndAltitudekeys and use it as the returned value.

File: transform/convert_quadkey_and_Vertical_id.go [hunk 3]
```diff
@@ -342,6 +348,86 @@ func ConvertExtendedSpatialIDsToQuadkeysAndVerticalIDs(extendedSpatialIDs []stri
...
+	extendedSpatialIDToQuadkeyAndVerticalID := []*object.FromExtendedSpatialIDToQuadkeyAndVerticalID{}
+
+	// validate zoom levels
+	if !quadkeyCheckZoom(outputHZoom, outputVZoom) {
+		return []*object.FromExtendedSpatialIDToQuadkeyAndVerticalID{}, errors.NewSpatialIdError(errors.InputValueErrorCode, "")
+	}
+	// outputのZoomレベルが指定されている前提のため、QuadkeyとAltitudeKeyのみを比較
+	deduplication := map[[2]int64]interface{}{}
```
Comment (HarutakaMatsumoto) [inline transform/convert_quadkey_and_Vertical_id.go:359]: This is a typo.
The correct spell is "duplicate".

**Review by HarutakaMatsumoto — CHANGES_REQUESTED** [review]
I reviewed it.
Check it out.

File: common/object/id_object.go [hunk 4]
```diff
@@ -1,6 +1,59 @@
...
+// FromExtendedSpatialIDToQuadkeyAndAltitudeKey 拡張空間IDから変換したquadkeyと標高のIDの組み合わせを管理する構造体
+type FromExtendedSpatialIDToQuadkeyAndAltitudeKey struct {
+	// quadkey zoom level / quadkeyの精度
+	quadkeyZoom int64
+	// a list of quadkey-altitude key sets [[quadkey,vIndex]...]
+	innerIDList [][2]int64
+	// vertical zoom level / 高さ方向の精度
+	vZoom int64
+	// altitude range scalar is s, where 2^25-s = altitude range (max altitude - min altitude)
+	altitudeRangeScalar int64
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:13]: Append verticalIndexOffset.

File: common/object/id_object.go [hunk 5]
```diff
@@ -1,6 +1,59 @@
...
 package object
 
+// FromExtendedSpatialIDToQuadkeyAndAltitudeKey 拡張空間IDから変換したquadkeyと標高のIDの組み合わせを管理する構造体
+type FromExtendedSpatialIDToQuadkeyAndAltitudeKey struct {
+	// quadkey zoom level / quadkeyの精度
+	quadkeyZoom int64
+	// a list of quadkey-altitude key sets [[quadkey,vIndex]...]
+	innerIDList [][2]int64
+	// vertical zoom level / 高さ方向の精度
+	vZoom int64
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:11]: Rename vZoom to altitudekeyZoom.
The same applies the others.

File: common/object/id_object.go [hunk 6]
```diff
@@ -1,6 +1,59 @@
 // 拡張空間IDパッケージ
 package object
 
+// FromExtendedSpatialIDToQuadkeyAndAltitudeKey 拡張空間IDから変換したquadkeyと標高のIDの組み合わせを管理する構造体
+type FromExtendedSpatialIDToQuadkeyAndAltitudeKey struct {
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:5]: Rename AltitudeKey to Altitudekey.
The same applies the others.

**Review by HarutakaMatsumoto — CHANGES_REQUESTED** [review]
I reviewed this.
Cope with the problems.

File: transform/convert_quadkey_and_Vertical_id.go (hunk 1 above)
Comment (buscandoaverroes) [reply transform/convert_quadkey_and_Vertical_id.go:681]: altitudeRangeScalar scales the altitudeRange by a factor of 2^n, whereas indexOffset shifts the altitude range up or down linearly

File: transform/convert_quadkey_and_Vertical_id.go [hunk 7]
```diff
@@ -576,6 +661,134 @@ func convertVerticallIDToBit(vZoom int64, vIndex int64, outputZoom int64, maxHei
...
+		return nil, error
+	}
+
+	// Determine the vertical index/indices to return.
+	// a) always return the lowerBound index. Regardless of the difference between the inputZoom and outputZoom,
+	// mathematically the altitude associated with the lower bounds will always satisfy the solution set.
+	// b) cycle through indices from lowerBounds+1 to upperBounds with i. If the minAltitude associated
+	// with i is less than the maxAltitude of the inputIndex, then the altitude associated with i
+	// satisfies the solution set. Add this index to the list of indices to return
+	outputIndexes = append(outputIndexes, lowerBound)
```
[the remark this answers is not in the document]
Comment (HarutakaMatsumoto) [reply transform/convert_quadkey_and_Vertical_id.go:704]: Thank you for your improvement!

File: transform/convert_quadkey_and_Vertical_id.go (hunk 1 above)
Comment (HarutakaMatsumoto) [reply transform/convert_quadkey_and_Vertical_id.go:682]: altitudeRange of SpatialIDs is 33554432 - -33554432 = 2^26, right?
So altitudeRange of QuadkeysAndAltitudekeys is 15872 - -512=2^14.
But you set 26 - 14 = 12 ≒ 11 as altitudeRangeScalar.
I think this is not altitudeRangeScalar but zoomOffset.

Comment (HarutakaMatsumoto) [reply transform/convert_quadkey_and_Vertical_id.go:682]: MEMO: @buscandoaverroes rejects the proposal by @HarutakaMatsumoto.

File: transform/convert_quadkey_and_Vertical_id_test.go [hunk 8]
```diff
@@ -597,3 +689,83 @@ func TestSpatialIDCheckZoom(t *testing.T) {
...
+		{inputZoom: 25, outputZoom: 27, inputIndex: 100, altitudeRangeScalar: 1, indexOffset: 0, expectedOutputIndex: []int64{800}},
+	}
+
+	for _, p := range datas {
+		result, error := convertVerticalIndex(p.inputIndex, p.inputZoom, p.outputZoom, p.altitudeRangeScalar, p.indexOffset)
+		if error != nil {
+			t.Log(t.Name())
+			t.Error(error)
+		}
+		for i := 0; i < len(p.expectedOutputIndex); i++ {
```
Comment (HarutakaMatsumoto) [inline transform/convert_quadkey_and_Vertical_id_test.go:728]: This is wrong in the case where len(p.expectedOutpuIndex) is less than len(result).
Indeed, the datas[10] is not checked enough.
Append length check.

File: transform/convert_quadkey_and_Vertical_id.go [hunk 9]
```diff
@@ -342,6 +342,86 @@ func ConvertExtendedSpatialIDsToQuadkeysAndVerticalIDs(extendedSpatialIDs []stri
...
 
+func ConvertExtendedSpatialIDsToQuadkeysAndAltitudekeys(extendedSpatialIDs []string, outputHZoom int64, outputVZoom int64, altitudeRangeScalar int64, verticalIndexOffset int64) ([]*object.FromExtendedSpatialIDToQuadkeyAndAltitudekey, error) {
+	extendedSpatialIDToQuadkeyAndAltitudekey := []*object.FromExtendedSpatialIDToQuadkeyAndAltitudekey{}
+
+	// validate zoom levels
+	if !quadkeyCheckZoom(outputHZoom, outputVZoom) {
+		return []*object.FromExtendedSpatialIDToQuadkeyAndAltitudekey{}, errors.NewSpatialIdError(errors.InputValueErrorCode, "")
+	}
+	// outputのZoomレベルが指定されている前提のため、QuadkeyとAltitudekeyのみを比較
+	duplicate := map[[2]int64]interface{}{}
```
Comment (HarutakaMatsumoto) [inline transform/convert_quadkey_and_Vertical_id.go:353]: I am sorry, there is a word [deduplication](https://ejje.weblio.jp/content/deduplication).
Please revert it.

File: common/object/id_object.go [hunk 10]
```diff
@@ -1,6 +1,70 @@
 // 拡張空間IDパッケージ
 package object
 
+// FromExtendedSpatialIDToQuadkeyAndAltitudekey 拡張空間IDから変換したquadkeyと標高のIDの組み合わせを管理する構造体
+type FromExtendedSpatialIDToQuadkeyAndAltitudekey struct {
+	// quadkey zoom level / quadkeyの精度
+	quadkeyZoom int64
+	// a list of quadkey-altitude key sets [[quadkey,vIndex]...]
+	innerIDList [][2]int64
+	// vertical zoom level / 高さ方向の精度
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:10]: This is altitudekey zoom level.
Update the explanation.

File: common/object/id_object.go [hunk 11]
```diff
@@ -1,6 +1,70 @@
...
+	altitudekeyZoom int64
+	// altitude range scalar is s, where 2^25-s = altitude range (max altitude - min altitude)
+	altitudeRangeScalar int64
+	// the difference in index terms of min and max altitude from the standard range (0, 2^25 meters)
+	verticalIndexOffset int64
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:17]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 12]
```diff
@@ -1,6 +1,70 @@
...
+func NewFromExtendedSpatialIDToQuadkeyAndAltitudekey(quadkeyZoom int64, innerIDList [][2]int64, altitudekeyZoom int64, altitudeRangeScalar int64, verticalIndexOffset int64) *FromExtendedSpatialIDToQuadkeyAndAltitudekey {
+	a := &FromExtendedSpatialIDToQuadkeyAndAltitudekey{}
+	a.SetQuadkeyZoom(quadkeyZoom)
+	a.SetInnerIDList(innerIDList)
+	a.SetVerticalZoom(altitudekeyZoom)
+	a.SetAltitudeRangeScalar(altitudeRangeScalar)
+	a.SetVerticalIndexOffset(verticalIndexOffset)
+	return a
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:27]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 13]
```diff
@@ -1,6 +1,70 @@
...
+func (a *FromExtendedSpatialIDToQuadkeyAndAltitudekey) SetQuadkeyZoom(quadkeyZoom int64) {
+	a.quadkeyZoom = quadkeyZoom
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:31]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 14]
```diff
@@ -1,6 +1,70 @@
...
+func (a *FromExtendedSpatialIDToQuadkeyAndAltitudekey) SetInnerIDList(innerIDList [][2]int64) {
+	a.innerIDList = innerIDList
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:35]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 15]
```diff
@@ -1,6 +1,70 @@
...
+func (a *FromExtendedSpatialIDToQuadkeyAndAltitudekey) SetVerticalZoom(altitudekeyZoom int64) {
+	a.altitudekeyZoom = altitudekeyZoom
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:39]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 16]
```diff
@@ -1,6 +1,70 @@
...
+func (a *FromExtendedSpatialIDToQuadkeyAndAltitudekey) SetAltitudeRangeScalar(aScalar int64) {
+	a.altitudeRangeScalar = aScalar
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:43]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 17]
```diff
@@ -1,6 +1,70 @@
...
+func (a *FromExtendedSpatialIDToQuadkeyAndAltitudekey) SetVerticalIndexOffset(verticalIndexOffset int64) {
+	a.verticalIndexOffset = verticalIndexOffset
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:47]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 18]
```diff
@@ -1,6 +1,70 @@
...
+func (a *FromExtendedSpatialIDToQuadkeyAndAltitudekey) QuadkeyZoom() int64 {
+	return a.quadkeyZoom
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:51]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 19]
```diff
@@ -1,6 +1,70 @@
...
+func (a *FromExtendedSpatialIDToQuadkeyAndAltitudekey) InnerIDList() [][2]int64 {
+	return a.innerIDList
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:55]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 20]
```diff
@@ -1,6 +1,70 @@
...
+func (a *FromExtendedSpatialIDToQuadkeyAndAltitudekey) AltitudekeyZoom() int64 {
+	return a.altitudekeyZoom
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:59]: This is a public function.
So append the explanation.

File: common/object/id_object.go [hunk 21]
```diff
@@ -1,6 +1,70 @@
...
+func (a *FromExtendedSpatialIDToQuadkeyAndAltitudekey) AltitudeRangeScalar() int64 {
+	return a.altitudeRangeScalar
+}
+
```
Comment (HarutakaMatsumoto) [inline common/object/id_object.go:63]: This is a public function.
So append the explanation.

**Review by HarutakaMatsumoto — CHANGES_REQUESTED** [review]
I reviewed this.
Cope with the problems.

File: transform/convert_quadkey_and_Vertical_id.go [hunk 22]
```diff
@@ -576,6 +656,96 @@ func convertVerticallIDToBit(vZoom int64, vIndex int64, outputZoom int64, maxHei
...
+
+	// check to make sure the input index exists in the input world
+	inputResolution := calculateArithmaticShift(1, inputZoom)
+
+	maxInputIndex := int64(inputResolution - 1)
+	minInputIndex := int64(0)
+
+	outputResolution := calculateArithmaticShift(1, outputZoom)
+	maxOutputIndex := int64(outputResolution-1) + indexOffset
+	minOutputIndex := indexOffset
```
[the remark this answers is not in the document]
Comment (buscandoaverroes) [reply transform/convert_quadkey_and_Vertical_id.go:728]: minOutputIndex can be negative if indexOffset is negative.

File: transform/convert_quadkey_and_Vertical_id.go [hunk 23]
```diff
@@ -576,6 +664,97 @@ func convertVerticallIDToBit(vZoom int64, vIndex int64, outputZoom int64, maxHei
...
+}
+
+// converts a vertical index from one set of zoom parameters to another disregarding the floor() cacluation. This creates a simplier system of equations where the solu	tion set for height is a single variable. However, this does not describe the full solution set of height since we have excluded the floor calculation; it describes the condition where m = x, given m = floor(x) if and only if m <= x < m +1;
+func calculateMinVerticalIndex(inputIndex int64, inputZoom int64, outputZoom int64, altitudeRangeScalar int64, indexOffset int64) (int64, error) {
+
+	// check to make sure the input index exists in the input world
+	inputResolution := CalculateArithmeticShift(1, inputZoom)
+
+	maxInputIndex := int64(inputResolution - 1)
+	minInputIndex := -int64(inputResolution + 1)
```
Comment (HarutakaMatsumoto) [inline transform/convert_quadkey_and_Vertical_id.go:724]: If you set inputZoom as 0, this minInputIndex is -2.
But the correct minInputIndex is -1 as 18 page of [the specification]( https://www.ipa.go.jp/digital/architecture/Individual-link/nq6ept000000g0fh-att/4dspatio-temporal-guideline-gamma.pdf).
And inputResolution is int64, so the type conversion is  wasteful.
Modify this to `	minInputIndex := -inputResolution`.
The same applies the others.

File: transform/convert_quadkey_and_Vertical_id.go [hunk 24]
```diff
@@ -576,6 +664,97 @@ func convertVerticallIDToBit(vZoom int64, vIndex int64, outputZoom int64, maxHei
...
+		return 0, errors.NewSpatialIdError(errors.InputValueErrorCode, "output index does not exist with given outputZoom, altitudeRangeScalar, and indexOffset")
+	}
+
+	return outputIndex, nil
+
+}
+
+// computes arithmatic shift of index and shift parameters. When index = 1, similar to returning 2^shift. When index > 1,
+// similar to returning index*2^shift.
+func CalculateArithmeticShift(index int64, shift int64) int64 {
```
Comment (HarutakaMatsumoto) [inline transform/convert_quadkey_and_Vertical_id.go:747]: Why did you make only this function public?
You can test private functions.
If it is not necessary, revert it for reducing the maintenance costs and the learning costs.

**Review by HarutakaMatsumoto — CHANGES_REQUESTED** [review]
I reviewed it.
Check it out.

**Pull request merged**